peps/pep-0747.rst

```diff
@@ -0,0 +1,856 @@
+
+``TypeExpr[]`` is covariant in its argument type, just like ``type[]``:
+
+-  ``TypeExpr[T1]`` is a subtype of ``TypeExpr[T2]`` iff ``T1`` is a
+   subtype of ``T2``.
+-  ``type[C1]`` is a subtype of ``TypeExpr[C2]`` iff ``C1`` is a subtype
+   of ``C2``.
+
+A plain ``type`` can be assigned to a plain ``TypeExpr`` but not the
+other way around:
+
+-  ``type[Any]`` is consistent with ``TypeExpr[Any]``. (But not the
+   other way around.)
+
+``TypeExpr[]`` is a kind of object, just like ``type[]``:
+
+-  ``TypeExpr[T]`` for any ``T`` is a subtype of ``object``.
+
+``TypeExpr[T]``, where ``T`` is a type variable, is assumed to have all
+the attributes and methods of ``object`` and is not callable.
+
+
+.. _interactions_with_type_variables:
+
+Interactions with type variables
+--------------------------------
+
+``TypeExpr[]`` can be parameterized by a type variable that is used elsewhere within
+the same function definition:
+
+::
+
+   def as_instance[T](typx: TypeExpr[T]) -> T | None:
+       return typx() if isinstance(typx, type) else None
+
+
+Interactions with type[]
+------------------------
+
+Both ``TypeExpr[]`` and ``type[]`` can be parameterized by the same type
+variable within the same function definition:
+
+::
+
+   def as_type[T](typx: TypeExpr[T]) -> type[T] | None:
+       return typx if isinstance(typx, type) else None
+
+
+Interactions with TypeIs[] and TypeGuard[]
+------------------------------------------
+
+A type variable parameterizing a ``TypeExpr[]`` can also be used by a ``TypeIs[]``
+within the same function definition:
+
+::
+
+   def isassignable[T](value: object, typx: TypeExpr[T]) -> TypeIs[T]: ...
+
+   count: int | str = ...
+   if isassignable(count, int):
+       assert_type(count, int)
+   else:
+       assert_type(count, str)
+
+or by a ``TypeGuard[]`` within the same function definition:
+
+::
+
+   def isdefault[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]:
+       return (value == typx()) if isinstance(typx, type) else False
+
+   value: int | str = ''
+   if isdefault(value, int):
+       assert_type(value, int)
+       assert 0 == value
+   elif isdefault(value, str):
+       assert_type(value, str)
+       assert '' == value
+   else:
+       assert_type(value, int | str)
+
+
+.. _interactions_with_annotated:
+
+Interactions with Annotated[]
+-----------------------------
+
+``Annotated[]`` forms preserve their metadata at runtime:
+
+::
+
+   >>> ValueRange: TypeAlias = slice
+   >>> PositiveInt: TypeAlias = Annotated[int, ValueRange(1, float('inf'))]
+   >>> PositiveInt
+   typing.Annotated[int, slice(1, inf, None)]  # NOT: int
+
+However ``Annotated[]`` forms need not preserve their metadata at
+typechecking-time:
+
+::
+
+   count: PositiveInt = 1
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+
+In particular when an ``Annotated[]`` argument is passed to a ``TypeExpr[]``
+parameterized by a type variable that is also used by a ``TypeIs[]``
```

Comment:
Hm, this actually becomes problematic with `TypeIs`, because whether `Annotated` is preserved is going to affect whether the `else` type still includes `int` in your example. I think it's OK to leave that unspecified though.

Reply:
(This comment applies to the "Interactions with Annotated[]" section, which has been deleted entirely from the latest draft.)